typed-racket-lib/typed-racket/typecheck/tc-structs.rkt

```diff
@@ -367,24 +367,32 @@
   (define constructor-binding
     (make-def-binding (struct-names-constructor names)
                       (poly-wrapper (->* all-fields poly-base))))
+
+  (define extra-constructor-binding
+    (if extra-constructor
+        (make-def-binding extra-constructor
+                          (poly-wrapper (->* all-fields poly-base)))
+        #f))
   (define constructor-bindings
     (cons constructor-binding
           (if extra-constructor
-              (list (make-def-binding extra-constructor
-                                      (poly-wrapper (->* all-fields poly-base))))
+              (list extra-constructor-binding)
               null)))
 
   (for ([b (in-list (append constructor-bindings bindings))])
     (register-type (binding-name b) (def-binding-ty b)))
-
+    
   (append
-    (if (free-identifier=? (struct-names-type-name names)
-                           (struct-names-constructor names))
+   (if (free-identifier=? (struct-names-type-name names)
+                          (struct-names-constructor names))
       null
       (list constructor-binding))
    (cons
-     (make-def-struct-stx-binding (struct-names-type-name names) si (def-binding-ty constructor-binding))
-     bindings)))
+    (make-def-struct-stx-binding (struct-names-type-name names) si (def-binding-ty constructor-binding))
+    (if extra-constructor
+        (cons (make-def-struct-stx-binding extra-constructor si (def-binding-ty extra-constructor-binding))
```

Comment:
we could also use that type here instead of extracting it from the `def-binding`.

Reply:
Yes. Also I don't think using `(make-def-struct-stx-binding extra-constructor ...)` makes sense here, because the field of def-struct-stx-binding is the name of the structure.  So I rewrote my changes